packages/nuxt/src/pages/module.ts

```diff
@@ -228,30 +258,77 @@ export default defineNuxtModule({
         },
       }
 
-      nuxt.hook('prepare:types', ({ references }) => {
+      nuxt.hook('prepare:types', ({ references, tsConfig }) => {
         // This file will be generated by unplugin-vue-router
-        references.push({ path: declarationFile })
+        references.push({ path: typedRouterDtsPathRelative })
+        // This file will be generated by us
+        references.push({ path: fileRouteNameMapDtsPathRelative })
         references.push({ types: 'unplugin-vue-router/client' })
+
+        // TODO: right now we add the VLS plugin whenever `typedPages: true` and `languageFeatures: true`,
+        // but we probably want to add the plugin only when `languageFeatures: true` and then set some flags
+        // to `true` when `typedPages: true`.
+        if (nuxt.options.experimental.languageFeatures) {
+          const pluginPath = import.meta.resolve('@nuxt/language-features').replace(/^file:\/\//, '')
+          const pluginPathRelative = relative(nuxt.options.buildDir, pluginPath)
+
+          // TODO: tried a bunch of these, but can't resolve to a subdep with `resolveModulePath` (nuxt -> @nuxt/language-features)
+          // const nuxtModulePath = dirname(resolveModulePath('nuxt/package.json'))
+          // console.log(nuxtModulePath)
+          // console.log(resolveModulePath('@nuxt/language-features', { from: nuxtModulePath }))
+          // const p = relative(nuxt.options.buildDir, resolveModulePath('@nuxt/language-features'))
+
+          tsConfig.vueCompilerOptions ??= {}
+          tsConfig.vueCompilerOptions.plugins ??= []
+          tsConfig.vueCompilerOptions.plugins.push(pluginPathRelative)
+        }
       })
 
       const context = createRoutesContext(resolveOptions(typedRouterOptions))
-      const dtsFile = resolve(nuxt.options.buildDir, declarationFile)
-      await mkdir(dirname(dtsFile), { recursive: true })
-      await context.scanPages(false)
+
+      // TODO: ensure this function doesn't run multiple times in parallel. if it does, stop the previous run.
+      async function generateTypedRouterFiles (startWatchers: boolean) {
+        // ensure types directory exists
+        await mkdir(dirname(typedRouterDtsPathAbsolute), { recursive: true })
+
+        await context.scanPages(startWatchers)
+
+        const fileRouteNameMap = Array.from(addedPagePaths.values())
```

Comment:
Not all routes in the `basic-types` test are included. Figure out why. Maybe `pages:extend` related, but most likely because we call `addedPagePaths.clear()` at the wrong time.